File: solution/1000-1099/1048.Longest String Chain/Solution.cpp
class Solution {
public:
    int longestStrChain(vector<string>& words) {
        sort(words.begin(), words.end(), [](auto& a, auto& b) { return a.size() < b.size(); });
        int ans = 0;
        unordered_map<string, int> d;
        for (auto& s : words) {
            int x = 1;
            for (int i = 0; i < s.size(); ++i) {
                string t = s.substr(0, i) + s.substr(i + 1);
                x = max(x, d[t] + 1);
            }
            d[s] = x;
            ans = max(ans, x);
        }
        return ans;
    }
};

File: solution/1000-1099/1048.Longest String Chain/Solution.java
class Solution {
    public int longestStrChain(String[] words) {
        Arrays.sort(words, Comparator.comparingInt(String::length));
        Map<String, Integer> d = new HashMap<>();
        int ans = 0;
        for (String s : words) {
            int x = 0;
            for (int i = 0; i < s.length(); ++i) {
                String t = s.substring(0, i) + s.substring(i + 1);
                x = Math.max(x, d.getOrDefault(t, 0) + 1);
            }
            d.put(s, x);
            ans = Math.max(ans, x);
        }
        return ans;
    }
}